File: desktopclient/src/main/java/com/goexp/galgame/gui/view/task/GameCharListTask.java
package com.goexp.galgame.gui.view.task;

import com.goexp.galgame.gui.db.mongo.query.GameQuery;
import com.goexp.galgame.gui.model.Game;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.concurrent.Task;

import static com.mongodb.client.model.Filters.eq;
import static com.mongodb.client.model.Projections.include;

public class GameCharListTask extends Task<ObservableList<Game.GameCharacter>> {

    private int gameId;

    private GameQuery.GameCharQuery gameCharQuery = new GameQuery.GameCharQuery();

    public GameCharListTask(int gameId) {
        this.gameId = gameId;

    }

    @Override
    protected ObservableList<Game.GameCharacter> call() {


        var g = GameQuery.GameCharQuery.tlp.query()
                .where(eq(gameId))
                .select(include("gamechar"))
                .one();

        return FXCollections.observableArrayList(g.gameCharacters);
    }
}

File: desktopclient/src/main/scala/com/goexp/galgame/gui/task/game/search/sub/GameCharListTask.scala
package com.goexp.galgame.gui.task.game.search.sub

import com.goexp.common.util.string.Strings
import com.goexp.galgame.common.model.game.GameCharacter
import com.goexp.galgame.gui.db.mongo.query.GameQuery
import com.mongodb.client.model.Filters
import javafx.collections.{FXCollections, ObservableList}
import javafx.concurrent.Task

import scala.jdk.CollectionConverters._

class GameCharListTask(private[this] val gameId: Int) extends Task[ObservableList[GameCharacter]] {
  override protected def call: ObservableList[GameCharacter] = {
    GameQuery.personTlp.where(Filters.eq(gameId)).one()
      .map {
        _.gameCharacters
      }.map(persons => {
      persons.asScala.to(LazyList)
        .groupBy(p => {
          if (Strings.isEmpty(p.cv)) {
            if (Strings.isEmpty(p.img)) 3 else 2
          } else 1
        }).to(LazyList)
        .sortBy({ case (k, _) => k })
        .flatMap({ case (_, v) => v }).asJava
    })
      .map(FXCollections.observableArrayList(_))
      .getOrElse(FXCollections.emptyObservableList[GameCharacter])

  }
}